### PR TITLE
[clang-tidy][NFC][doc] Improve documentation for modernize-use-equals…

### LINKED ISSUE #65221: Unclear documentation of modernize-use-equals-delete
The docs of "modernize-use-equals-delete" don't quite explain what the purpose of the check is. What problem is it solving? Why is the new code preferred over the old code? Any references that one can read about it?

The closest I can find is a C++ Core Guideline that says that one should not mark a function as private in order to delete it - the modern way is to move the function to  *public* and mark it as deleted. 

However modernize-use-equals-delete does not do that - it doesn't move the deleted function from private to public, which is a bit confusing.

### PR DESCRIPTION
**Pull request by carlosgalvezp**
…-delete

So the purpose of the check is more clear. Update examples code to show compliant code.

Fixes #65221

### CODE REVIEW

File: clang-tools-extra/docs/clang-tidy/checks/modernize/use-equals-delete.rst [hunk 1]
```diff
@@ -3,22 +3,34 @@
...
 ===========================
 
+Prior to C++11, the only way to "delete" a given function was to make it
+``private`` and without definition, to generate a compiler error (calling
+private function) or a linker error (undefined reference).
+
+After C++11, the more idiomatic way to achieve this is by marking the functions
+as ``= delete``, and keeping them in the ``public`` section.
+
 This check marks unimplemented private special member functions with ``= delete``.
```
Comment (PiotrZSL) [inline clang-tools-extra/docs/clang-tidy/checks/modernize/use-equals-delete.rst:13]: This is duplication. Saying twice about function marked as delete.

File: clang-tools-extra/docs/clang-tidy/checks/modernize/use-equals-delete.rst [hunk 2]
```diff
@@ -3,22 +3,34 @@
 modernize-use-equals-delete
 ===========================
 
+Prior to C++11, the only way to "delete" a given function was to make it
```
Comment (PiotrZSL) [inline clang-tools-extra/docs/clang-tidy/checks/modernize/use-equals-delete.rst:6]: Begining of the description should say what check does, not a why.

**Review by PiotrZSL — COMMENTED** [review]
Maybe:

```
Identifies unimplemented private special member functions, recommends using '= delete' for them, and suggests relocating such deleted functions from the private to the public section.

Before the introduction of C++11, the primary method to effectively "erase" a particular function involved declaring it as private without providing a definition. This approach would result in either a compiler error (when attempting to call a private function) or a linker error (due to an undefined reference).

However, subsequent to the advent of C++11, a more conventional approach emerged for achieving this purpose. It involves flagging functions as "= delete" and retaining them within the public section of the code.

To prevent false positives, this check is only applicable within a translation unit where all other member functions have been implemented. The check will generate partial fixes by introducing "= delete," but the user is responsible for manually relocating it to the public section.
```

File: clang-tools-extra/docs/clang-tidy/checks/modernize/use-equals-delete.rst [hunk 3]
```diff
@@ -3,22 +3,37 @@
...
-  // becomes
-
-  struct A {
-  private:
+  // Example: good
+  class A {
+   public:
     A(const A&) = delete;
     A& operator=(const A&) = delete;
   };
```
Comment (PiotrZSL) [inline clang-tools-extra/docs/clang-tidy/checks/modernize/use-equals-delete.rst:39]: This check got option `IgnoreMacros`  (local or global), would be good to document it. It's a basic one, but is...

Comment (carlosgalvezp) [reply clang-tools-extra/docs/clang-tidy/checks/modernize/use-equals-delete.rst:39]: It's already documented on line 42, is there anything else you would like to add?

**Comment by PiotrZSL** [comment]
Other interesting thing is that currently this check is limited just to LangOpts.CPlusPlus, when it should be to LangOpts.CPlusPlus11. And in such case documentation could mention that requires C++11 or later.

File: clang-tools-extra/docs/clang-tidy/checks/modernize/use-equals-delete.rst (hunk 3 above)
Comment (PiotrZSL) [reply clang-tools-extra/docs/clang-tidy/checks/modernize/use-equals-delete.rst:39]: Ach, didnt see it, had to click button to load rest of diff, then it's fine...

**Comment by carlosgalvezp** [comment]
Yep, we could certainly change it to C++11 or later, but I think it's out of the scope for this NFC patch. Regarding documentation, I haven't seen that we document this detail in other checks, it seems like an implementation detail that could cause maintenance effort if documented.

**Review by PiotrZSL — APPROVED** [review]
LGTM

**Comment by PiotrZSL** [comment]
Do we want to sync comment above check class in code with first sentence in documentation ?

**Comment by carlosgalvezp** [comment]
Yes good point, will fix!

File: clang-tools-extra/clang-tidy/modernize/UseEqualsDeleteCheck.h [hunk 4]
```diff
@@ -13,22 +13,9 @@
...
-/// \code
-///   struct A {
-///   private:
-///     A(const A&) = delete;
-///     A& operator=(const A&) = delete;
-///   };
-/// \endcode
+/// Identifies unimplemented private special member functions, and recommends
+/// using ``= delete`` for them, as well as relocating them from the ``private``
+/// to the ``public`` section.
```
Comment (PiotrZSL) [inline clang-tools-extra/clang-tidy/modernize/UseEqualsDeleteCheck.h:18]: looks like check currently does not limit private -> public to only special member functions, in such case we could write:
`as well as relocating deleted functions from the ``private`` to the ``public`` section.`

**Review by PiotrZSL — APPROVED** [review]
No more comments from me.

**Pull request merged**